includes/class-integrity-check-utils.php

```diff
@@ -0,0 +1,190 @@
+	 */
+	public static function get_membership_data( $max_records = null ) {
+		if ( ! class_exists( 'WC_Memberships_User_Membership' ) ) {
+			return [];
+		}
+
+		global $wpdb;
+
+		// phpcs:disable WordPressVIPMinimum.Variables.RestrictedVariables.user_meta__wpdb__users
+		$query = "
+			SELECT 
+				u.user_email,
+				p.post_status as status,
+				pm_network.meta_value as network_id
+			FROM {$wpdb->posts} p
+			INNER JOIN {$wpdb->users} u ON p.post_author = u.ID
+			INNER JOIN {$wpdb->postmeta} pm_network ON p.post_parent = pm_network.post_id AND pm_network.meta_key = %s
+			INNER JOIN (
+				SELECT 
+					p2.post_author,
+					pm2.meta_value,
+					MAX(p2.post_date) as max_date
+				FROM {$wpdb->posts} p2
+				INNER JOIN {$wpdb->postmeta} pm2 ON p2.post_parent = pm2.post_id AND pm2.meta_key = %s
+				WHERE p2.post_type = 'wc_user_membership'
+				AND pm2.meta_value IS NOT NULL
+				AND pm2.meta_value != ''
+				GROUP BY p2.post_author, pm2.meta_value
+			) latest ON p.post_author = latest.post_author 
+				AND pm_network.meta_value = latest.meta_value 
+				AND p.post_date = latest.max_date
+			WHERE p.post_type = 'wc_user_membership'
+			AND pm_network.meta_value IS NOT NULL
+			AND pm_network.meta_value != ''
+			ORDER BY LOWER(u.user_email) ASC
+		";
+		// phpcs:enable WordPressVIPMinimum.Variables.RestrictedVariables.user_meta__wpdb__users
+
+		if ( $max_records ) {
+			$query .= $wpdb->prepare( ' LIMIT %d', $max_records );
+		}
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery,WordPress.DB.DirectDatabaseQuery.NoCaching,WordPressVIPMinimum.Variables.RestrictedVariables.user_meta__wpdb__users,WordPress.DB.PreparedSQL.NotPrepared
+		$results = $wpdb->get_results( $wpdb->prepare( $query, Memberships_Admin::NETWORK_ID_META_KEY, Memberships_Admin::NETWORK_ID_META_KEY ) );
+
+		$membership_data = [];
+		foreach ( $results as $result ) {
+			$membership_data[] = [
+				'email'      => strtolower( $result->user_email ),
+				'status'     => $result->status,
+				'network_id' => $result->network_id,
+			];
+		}
+
+		return $membership_data;
+	}
+
+	/**
+	 * Get membership data within an email range
+	 *
+	 * Filters memberships by email address range rather than positional offset.
+	 * This enables range-based chunking that's resilient to data shifts when
+	 * memberships are added/removed from the beginning or middle of the dataset.
+	 *
+	 * @param string   $start_email The start email (inclusive).
+	 * @param string   $end_email The end email (inclusive).
+	 * @param int|null $max_records Maximum number of records to return (for testing).
+	 * @return array Array of (email, status, network_id) data
+	 */
+	public static function get_membership_data_range( $start_email, $end_email, $max_records = null ) {
+		if ( ! class_exists( 'WC_Memberships_User_Membership' ) ) {
+			return [];
+		}
+
+		global $wpdb;
+
+		// phpcs:disable WordPressVIPMinimum.Variables.RestrictedVariables.user_meta__wpdb__users
+		$query = "
```

Comment:
Non-blocking: Not sure if its worth it, but this query and the one above look to be almost identical aside from 2 lines. Maybe its worth moving this into its own helper method so we aren't duplicating so much?

Reply:
Good call, refactored in recent changes